### PR TITLE
Introduce tf_http_archive

### PR DESCRIPTION
**Pull request by jart**
I decided to give https://github.com/tensorflow/tensorflow/pull/14813 another try.

### CODE REVIEW

File: third_party/repo.bzl [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+tf_http_archive = repository_rule(
+    implementation=_tf_http_archive,
+    attrs={
+        "sha256": attr.string(mandatory=True),
+        "urls": attr.string_list(mandatory=True, allow_empty=False),
+        "strip_prefix": attr.string(),
+        "type": attr.string(),
+        "delete": attr.string_list(),
+        "patch_file": attr.label(),
+        "build_file": attr.label(),
```
Comment (gunan) [inline third_party/repo.bzl]: Would it be too confusing if we added a  `bind` arg, which is a list of pairs or a dict, then if it is set we also create all the `native.bind` rules we have in `tf_workspace.bzl`?

**Review by gunan — APPROVED** [review]

File: third_party/repo.bzl (hunk 1 above)
Comment (jart) [reply third_party/repo.bzl]: I documented the bind() statements instead, since we can't create them from a Skylark rule.

**Comment by jart** [comment]
PTAL

**Review by gunan — APPROVED** [review]

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
I see this error is breaking all the builds:
```
ERROR: /tmpfs/tmp/bazel/external/grpc/BUILD:539:1: no such target '//external:zlib': target 'zlib' not declared in package 'external' defined by /tmpfs/src/github/tensorflow/WORKSPACE and referenced by '@grpc//:grpc_base_c'
WARNING: /tmpfs/src/github/tensorflow/tensorflow/core/BUILD:1815:1: in includes attribute of cc_library rule //tensorflow/core:framework_headers_lib: '../../external/nsync/public' resolves to 'external/nsync/public' not below the relative path of its package 'tensorflow/core'. This will be an error in the future. Since this rule was created by the macro 'cc_header_only_library', the error might have been caused by the macro implementation in /tmpfs/src/github/tensorflow/tensorflow/tensorflow.bzl:1127:30
Analyzing: 503 targets (343 packages loaded)
```

Maybe we missed one of the bind statements?

**Comment by jart** [comment]
Added.

**Comment by gunan** [comment]
It looks like this time we have a problem with this script:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/makefile/download_dependencies.sh
It parses the workspace.bzl file to get the links to our dependencies.

**Comment by jart** [comment]
Third time's a charm. test this please

**Comment by jart** [comment]
Build failures now appear to be unrelated.

**Comment by gunan** [comment]
Yes, all failures are known issues.
Merging.

**Pull request merged**